### DISCUSSION TITLE
What's the recommended way to handle/recover from a failure to get a connection from the pool?

### DISCUSSION THREAD

**Discussion by thomasmost** [diesel-rs/diesel#3323] [Q&A]
I'm building an API using Rocket and Diesel, and I'm managing the DbPool using Rocket's State. A search request handler might look like this, then:

```rs
#[get("/search?<term>")]
pub fn general_privileged(
  db_pool: &State<Pool>,
  _key: PrivilegedApiKey,
  term: &str,
) -> Json<Vec<SearchResult>> {
  let dao1_connection = db_pool.get().unwrap();
  let dao2_connection = db_pool.get().unwrap();
  let company_dao = CompanyDao::new(dao1_connection);
  let user_dao = UserDao::new(dao2_connection);
  let results = lib::search::general_privileged::search(term, company_dao, user_dao);
  Json(results)
}
```

As you can see, I'm unwrapping the connections here, which is not a good practice. In the case of a panic, it takes the service a long time to recover, which degrades performance. Obviously I could just return a `503 status instead of panicking (better!) but that's still not a great experience to have an API that frequently cannot talk to the DB.

I could use some help understanding a few things here:
1. Under what circumstances should I *expect* a failure to retrieve a connection from the pool; i.e. can I plan for and avoid these situations?
2. What are the recommended best practices for handling such failures?
3. What are the recommended best practices for responding to and recovering from such failures?

Thanks in advance!

**Comment by thomasmost** [comment]
Sounds like r2d2 returns connections to the pool automatically when they fall out of scope

**Comment by weiznich** [comment]
I'm on mobile, so this is answer might be a bit short:

> Under what circumstances should I expect a failure to retrieve a connection from the pool; i.e. can I plan for and avoid these situations?

The pool should return an error if it cannot receive a valid connection in the configured time period from the pool. That default timeout is something like 10s, so that only happens if either your database server is unreachable, or you have some long running requests that dead lock each other by trying to get another connection while the pool is already exhausted.

> What are the recommended best practices for handling such failures?
> What are the recommended best practices for responding to and recovering from such failures?

The only case that you can somehow handle is that one with the dead locking. Thats normally a bug in your code and should be fixed there. Any other error case is likely something where you cannot do much, as the database has gone away or something like that. Returning a 503 error in such cases should be fine. You should be able to do that by using the `?` operator.
